### PR TITLE
feat(api): Return api service level in rate limit headers

### PR DESCRIPTION
**Pull request by rifont**
### What change does this PR introduce?
* Modifies the get-api-rate-limit-maximum use-case to return the apiServiceLevel for the organization
* Updates the throttler guard to specify the apiServiceLevel used for the request in the `RateLimit-Policy` header
* Introduces a constant `'custom'` api service level to define when a custom Environment rate limit value was used
* Add instrumentation for critical API path guards/interceptors

### Why was this change needed?
In the interest of transparency, the API service level should be returned to the API client to ascertain which service level their request was evaluated against. In cases where an Environment has a custom rate limit specified, this information should also be made available to the client.

The API service level used for rate limiting should also be available for observability capabilities to determine how each service level is utilizing the platform.

### Other information (Screenshots)
_Service Level in policy header of API response_
image

### CODE REVIEW

File: apps/api/src/app/rate-limiting/usecases/evaluate-api-rate-limit/evaluate-api-rate-limit.types.ts [hunk 1]
```diff
@@ -35,4 +35,8 @@ export type EvaluateApiRateLimitResponseDto = {
    * The cost of the request.
    */
   cost: number;
+  /**
+   * The API service level used to evaluate the request.
+   */
+  apiServiceLevel: string;
```
Comment (rifont) [inline apps/api/src/app/rate-limiting/usecases/evaluate-api-rate-limit/evaluate-api-rate-limit.types.ts:41]: Intentionally type-widening here, as the evaluation doesn't need to know about the specific enum typing of the api service level, it is simply proxying the value.

File: apps/api/src/app/rate-limiting/usecases/get-api-rate-limit-maximum/get-api-rate-limit-maximum.spec.ts [hunk 2]
```diff
@@ -112,39 +125,78 @@ describe('GetApiRateLimitMaximum', async () => {
       });
     });
 
-    it('should return default api rate limit for the organizations apiServiceLevel when apiServiceLevel IS set on organization', async () => {
+    describe('Organization DOES have api service level specified', () => {
```
Comment (rifont) [inline apps/api/src/app/rate-limiting/usecases/get-api-rate-limit-maximum/get-api-rate-limit-maximum.spec.ts:128]: Organizing the cases for Org DOES/DOES NOT have api service level into `describe` blocks, to reuse the `beforeEach` initializers.

File: apps/api/src/app/rate-limiting/usecases/get-api-rate-limit-maximum/get-api-rate-limit-maximum.usecase.ts [hunk 3]
```diff
@@ -48,7 +49,9 @@ export class GetApiRateLimitMaximum {
     }
 
     let apiRateLimits: IApiRateLimitMaximum;
+    let apiServiceLevel: ApiServiceLevel;
     if (environment.apiRateLimits) {
+      apiServiceLevel = CUSTOM_API_SERVICE_LEVEL;
```
Comment (rifont) [inline apps/api/src/app/rate-limiting/usecases/get-api-rate-limit-maximum/get-api-rate-limit-maximum.usecase.ts:54]: When we have a custom rate limit set on the environment, return a constant with value of `'custom'` to enable metrics reporting on use of custom rate limits, and return this info to API client.

File: apps/api/src/app/rate-limiting/guards/throttler.guard.ts [hunk 4]
```diff
@@ -88,7 +94,7 @@ export class ApiRateLimitInterceptor extends ThrottlerGuard implements NestInter
     // Return early if the current user agent should be ignored.
     if (Array.isArray(ignoreUserAgents)) {
       for (const pattern of ignoreUserAgents) {
-        if (pattern.test(req.headers[HttpRequestHeaderKeysEnum.AUTHORIZATION.toLowerCase()])) {
+        if (pattern.test(req.headers[HttpRequestHeaderKeysEnum.USER_AGENT.toLowerCase()])) {
```
Comment (rifont) [inline apps/api/src/app/rate-limiting/guards/throttler.guard.ts:97]: This was an error from a recent update when updating to use the header enum.

File: packages/application-generic/src/services/feature-flags.service.ts [hunk 5]
```diff
@@ -112,6 +113,7 @@ export class FeatureFlagsService {
    *  - Default value with the value provided by Novu's environment variable if set
    *  - Default value with the value provided by the hardcoded fallback
    */
+  @Instrument()
```
Comment (rifont) [inline packages/application-generic/src/services/feature-flags.service.ts:116]: Instrumenting the feature flag read time for more comprehensive observability.

File: packages/application-generic/src/services/auth/user.auth.guard.ts [hunk 6]
```diff
@@ -29,6 +31,13 @@ export class UserAuthGuard extends AuthGuard([
     super();
   }
 
+  @Instrument()
+  canActivate(
+    context: ExecutionContext
+  ): boolean | Promise<boolean> | Observable<boolean> {
+    return super.canActivate(context);
+  }
+
```
Comment (rifont) [inline packages/application-generic/src/services/auth/user.auth.guard.ts:40]: Instrumenting the AuthGuard for creating future Auth Observability Dashboards.

File: apps/api/src/app/shared/framework/idempotency.interceptor.ts [hunk 7]
```diff
@@ -34,6 +34,7 @@ enum ReqStatusEnum {
 export class IdempotencyInterceptor implements NestInterceptor {
   constructor(private readonly cacheService: CacheService) {}
 
+  @Instrument()
```
Comment (rifont) [inline apps/api/src/app/shared/framework/idempotency.interceptor.ts:37]: Instrumenting the IdempotencyInterceptor for creating Idempotency Observability Dashboards.

File: packages/application-generic/src/logging/index.ts [hunk 8]
```diff
@@ -131,6 +131,7 @@ export function createNestLoggingModuleOptions(
           organizationId: req?.user?.organizationId || null,
         },
         authScheme: req?.authScheme,
+        rateLimitPolicy: res?.rateLimitPolicy,
```
Comment (rifont) [inline packages/application-generic/src/logging/index.ts:134]: Adding the rate limit policy to the logs for use in observability, specifically Rate Limit utilization.

**Review by Cliftonz — APPROVED** [review]

**Pull request merged**